libmamba/src/core/transaction.cpp

```diff
@@ -825,208 +828,231 @@ namespace mamba
-                            ));
-                            repr.postfix.set_value(
-                                fmt::format("{:<25}", extract_bar->last_active_task())
-                            );
-                        }
-                        repr.current.set_value(fmt::format("{:>3}", extract_bar->current()));
-                        repr.separator.set_value("/");
-
-                        std::string total_str;
-                        if (extract_bar->total() == std::numeric_limits<std::size_t>::max())
-                        {
-                            total_str = "?";
-                        }
-                        else
-                        {
-                            total_str = std::to_string(extract_bar->total());
-                        }
-                        repr.total.set_value(fmt::format("{:>3}", total_str));
-                    }
-                );
+                std::packaged_task task{ [=] { it->run(); } };
+                extract_tasks.push_back(task.get_future());
+                MainExecutor::instance().schedule(std::move(task));
             }
-
-            pbar_manager.start();
-            pbar_manager.watch_print();
         }
 
-        bool downloaded = multi_dl.download(MAMBA_DOWNLOAD_FAILFAST | MAMBA_DOWNLOAD_SORT);
-        bool all_valid = true;
-
-        if (!downloaded)
+        bool trigger_download(
+            MultiDownloadRequest requests,
+            const Context& context,
+            DownloadOptions options,
+            PackageDownloadMonitor* monitor
+        )
         {
-            LOG_ERROR << "Download didn't finish!";
-            return false;
+            auto result = download(std::move(requests), context, options, monitor);
+            bool all_downloaded = std::accumulate(
+                result.begin(),
+                result.end(),
+                true,
+                [](bool acc, const auto& r) { return acc && r; }
+            );
+            return all_downloaded;
         }
-        // make sure that all targets have finished extracting
-        while (!is_sig_interrupted())
+
+        bool check_all_valid(const FetcherList& fetchers, const ExtractRequestList& extract_requests)
         {
-            bool all_finished = true;
-            for (const auto& t : targets)
+            bool all_valid = true;
+            for (auto [fit, eit] = std::tuple{ fetchers.begin(), extract_requests.begin() };
+                 eit != extract_requests.end();
+                 ++fit, ++eit)
             {
-                if (!t->finished())
+                PackageExtractRequest::Result res = eit->get_result();
+                if (!res.valid || !res.extracted)
                 {
-                    all_finished = false;
-                    break;
+                    fit->clear_cache();
+                    all_valid = false;
+                    // TODO: check if we can remove this
+                    throw std::runtime_error(
```

Comment:
Throwing here means we never return `false` so the return type doesnt help at all the caller.
I agree with the comment that this throw should be removed and let the caller decide what to do once they have checked.

Basically this function does not just check vaility, it also updates a cache, so maybe there is a need for 2 separate fnuctions.
The validity check could be done with `std::all_of` but then you would need extract requests to know their fetchers.
Not saying to improve now, but maybe add a comment to improve this function later?

Reply:
> Basically this function does not just check vaility, it also updates a cache, so maybe there is a need for 2 separate fnuctions.

That would mean iterating twice, but I agree, that would be cleaner. Or we could rename the method `clean_invalid_cache` ?

> The validity check could be done with std::all_of but then you would need extract requests to know their fetchers.

Actually the requests already know their fetchers, so we could do this now. This would allow to iterate on the extract requests only. My only concerns, if we don't split the method, or if we split it but still want to iterate on the extract requests only, is that PackageExtractRequest should expose the clean_cache metho to forward the call to the fetcher. A bit weird, but I can live with it.